File: src/moonlight/recorder.py
"""Wraps `playwright codegen` and converts the recorded session into YAML.

Uses ast.parse for the Python source so locators like
`page.get_by_role("button", name="Yes")` survive intact — earlier regex-based
parsing collapsed them into broken CSS selectors.
"""
from __future__ import annotations

import ast
import subprocess
import sys
import tempfile
from pathlib import Path

import yaml


# Locator-chain actions emitted by codegen.
_ACTIONS = {"click", "fill", "press", "select_option", "check", "uncheck", "set_input_files"}


def _literal(node: ast.AST):
    if isinstance(node, ast.Constant):
        return node.value
    return None


def _parse_locator_call(call: ast.Call) -> dict | None:
    """Turn `page.get_by_role("button", name="Yes")` into {role: button, name: Yes}.

    Also handles `page.locator("css")` → {selector: css}.
    """
    if not isinstance(call.func, ast.Attribute):
        return None
    if not (isinstance(call.func.value, ast.Name) and call.func.value.id == "page"):
        return None

    method = call.func.attr
    pos = [_literal(a) for a in call.args]
    kw = {k.arg: _literal(k.value) for k in call.keywords if k.arg}

    if method == "locator":
        return {"selector": pos[0] if pos else ""}

    if method.startswith("get_by_"):
        kind = method[len("get_by_"):]   # role, label, text, placeholder, test_id, alt_text, title
        out: dict = {kind: pos[0] if pos else ""}
        for k, v in kw.items():
            if v is not None:
                out[k] = v
        return out

    return None


def _parse_line(line: str) -> dict | None:
    """Convert one codegen statement into a YAML step (or None)."""
    line = line.strip()
    if not line.startswith("await "):
        return None
    expr_src = line[len("await "):].rstrip()
    try:
        tree = ast.parse(expr_src, mode="eval").body
    except SyntaxError:
        return None
    if not isinstance(tree, ast.Call):
        return None

    func = tree.func

    # Case 1: bare `page.goto("url")` (one call, no chain)
    if isinstance(func, ast.Attribute) and isinstance(func.value, ast.Name) and func.value.id == "page":
        if func.attr == "goto" and tree.args:
            return {"action": "goto", "url": _literal(tree.args[0])}
        return None

    # Case 2: chain like LOCATOR.<ACTION>(...)
    if not isinstance(func, ast.Attribute):
        return None
    action = func.attr
    if action not in _ACTIONS:
        return None
    action_args = [_literal(a) for a in tree.args]

    locator_node = func.value
    if not isinstance(locator_node, ast.Call):
        return None
    loc = _parse_locator_call(locator_node)
    if loc is None:
        return None

    step: dict = {"action": "click" if action == "click" else action, **loc}

    if action == "fill":
        step["value"] = action_args[0] if action_args else ""
    elif action == "press":
        step["key"] = action_args[0] if action_args else ""
    elif action == "select_option":
        step["action"] = "select"
        step["value"] = action_args[0] if action_args else ""

    return step


def _parse_codegen(py_source: str) -> tuple[str | None, list[dict]]:
    base_url: str | None = None
    steps: list[dict] = []
    placeholder_idx = 0

    def add_screenshot(hint: str) -> None:
        nonlocal placeholder_idx
        placeholder_idx += 1
        steps.append({
            "action": "screenshot",
            "name": f"{hint}-{placeholder_idx:02d}",
            "ai_check": {
                "prompt": "TODO: describe what should be visible here",
                "expect": "yes",
            },
        })

    for raw in py_source.splitlines():
        step = _parse_line(raw)
        if not step:
            continue
        if step["action"] == "goto":
            url = step["url"]
            if base_url is None:
                base_url = url
                steps.append({"action": "goto", "url": "/"})
            else:
                steps.append({"action": "goto", "url": url})
            add_screenshot("after-goto")
            continue
        steps.append(step)
        if step["action"] in {"click", "press"}:
            add_screenshot(f"after-{step['action']}")

    return base_url, steps


def record(
    url: str,
    scenario_name: str,
    scenarios_dir: Path,
    viewport: dict[str, int],
    auth_state_path: Path | None = None,
) -> Path:
    """Launch playwright codegen, then write scenarios/<name>.yaml. Returns the path."""
    scenarios_dir.mkdir(parents=True, exist_ok=True)
    out_yaml = scenarios_dir / f"{scenario_name}.yaml"

    with tempfile.NamedTemporaryFile("w", suffix=".py", delete=False) as tmp:
        tmp_path = Path(tmp.name)

    cmd = [
        sys.executable, "-m", "playwright", "codegen",
        "--target", "python-async",
        "--output", str(tmp_path),
        f"--viewport-size={viewport['width']},{viewport['height']}",
    ]
    if auth_state_path is not None:
        auth_state_path.parent.mkdir(parents=True, exist_ok=True)
        cmd += ["--save-storage", str(auth_state_path)]
    cmd.append(url)

    print(f"[recorder] launching: {' '.join(cmd)}")
    print("[recorder] click through your app (incl. login), then close the Chromium window.")
    subprocess.run(cmd, check=True)

    source = tmp_path.read_text()
    base, steps = _parse_codegen(source)
    if not base:
        raise RuntimeError("no goto() detected — recording produced no usable steps")

    scenario: dict = {
        "name": scenario_name,
        "url": base,
        "viewport": viewport,
        "steps": steps,
    }
    if auth_state_path is not None and auth_state_path.exists():
        try:
            rel = auth_state_path.relative_to(scenarios_dir.parent)
            scenario["storage_state"] = str(rel)
        except ValueError:
            scenario["storage_state"] = str(auth_state_path)

    out_yaml.write_text(yaml.safe_dump(scenario, sort_keys=False, default_flow_style=False))
    print(f"[recorder] wrote {out_yaml} — open it and fill in the TODO ai_check prompts.")
    return out_yaml

File: internal/recorder/recorder.go
// Package recorder shells out to `npx playwright codegen` for the actual
// browser recording, then parses the emitted Python source into a YAML
// scenario. Codegen is the official Playwright tool — we don't reinvent it.
//
// Requires Node.js + npm on the tester's machine. If npx is missing, returns
// a clear error pointing at the install instructions.
package recorder

import (
	"fmt"
	"os"
	"os/exec"
	"regexp"
	"strings"

	"github.com/krisk248/moonlight/internal/scenario"
)

// Options for a recording session.
type Options struct {
	URL          string
	ScenarioName string
	Viewport     scenario.Viewport
	OutputPath   string // where to write the .yaml
}

// Record launches `npx playwright codegen` (which opens a Chromium window on
// the operator's desktop), waits for them to close it, and persists the
// recorded session as a YAML scenario at opts.OutputPath.
func Record(opts Options) (*scenario.Scenario, error) {
	if _, err := exec.LookPath("npx"); err != nil {
		return nil, fmt.Errorf("npx not found. Install Node.js + npm, then retry. (https://nodejs.org/)")
	}
	if opts.Viewport.Width == 0 {
		opts.Viewport = scenario.Viewport{Width: 1280, Height: 720}
	}

	tmpFile, err := os.CreateTemp("", "moonlight-codegen-*.py")
	if err != nil {
		return nil, err
	}
	tmpFile.Close()
	defer os.Remove(tmpFile.Name())

	args := []string{
		"--yes",
		"playwright",
		"codegen",
		"--target", "python-async",
		"--output", tmpFile.Name(),
		fmt.Sprintf("--viewport-size=%d,%d", opts.Viewport.Width, opts.Viewport.Height),
		opts.URL,
	}
	cmd := exec.Command("npx", args...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	if err := cmd.Run(); err != nil {
		return nil, fmt.Errorf("codegen failed: %w", err)
	}

	source, err := os.ReadFile(tmpFile.Name())
	if err != nil {
		return nil, err
	}

	base, steps := parseCodegen(string(source))
	if base == "" {
		return nil, fmt.Errorf("no goto() detected in recording")
	}

	sc := &scenario.Scenario{
		Name:     opts.ScenarioName,
		URL:      base,
		Viewport: opts.Viewport,
		Steps:    steps,
	}
	if opts.OutputPath != "" {
		if err := scenario.Save(sc, opts.OutputPath); err != nil {
			return nil, fmt.Errorf("save yaml: %w", err)
		}
	}
	return sc, nil
}

// ---------- parser ---------------------------------------------------------

var (
	rGoto         = regexp.MustCompile(`await page\.goto\("([^"]+)"\)`)
	rRoleClick    = regexp.MustCompile(`await page\.get_by_role\("([^"]+)"(?:,\s*name="([^"]+)")?\)\.click\(\)`)
	rRoleFill     = regexp.MustCompile(`await page\.get_by_role\("([^"]+)"(?:,\s*name="([^"]+)")?\)\.fill\("([^"]*)"\)`)
	rRolePress    = regexp.MustCompile(`await page\.get_by_role\("([^"]+)"(?:,\s*name="([^"]+)")?\)\.press\("([^"]+)"\)`)
	rLabelFill    = regexp.MustCompile(`await page\.get_by_label\("([^"]+)"\)\.fill\("([^"]*)"\)`)
	rLabelClick   = regexp.MustCompile(`await page\.get_by_label\("([^"]+)"\)\.click\(\)`)
	rTextClick    = regexp.MustCompile(`await page\.get_by_text\("([^"]+)"\)\.click\(\)`)
	rPlaceFill    = regexp.MustCompile(`await page\.get_by_placeholder\("([^"]+)"\)\.fill\("([^"]*)"\)`)
	rLocClick     = regexp.MustCompile(`await page\.locator\("([^"]+)"\)\.click\(\)`)
	rLocFill      = regexp.MustCompile(`await page\.locator\("([^"]+)"\)\.fill\("([^"]*)"\)`)
	rLocSelect    = regexp.MustCompile(`await page\.locator\("([^"]+)"\)\.select_option\("([^"]+)"\)`)
)

func parseCodegen(source string) (baseURL string, steps []scenario.Step) {
	placeholderIdx := 0
	addScreenshot := func(hint string) {
		placeholderIdx++
		steps = append(steps, scenario.Step{
			Action: "screenshot",
			Name:   fmt.Sprintf("%s-%02d", hint, placeholderIdx),
			DOMCheck: &scenario.DOMCheck{
				URLContains: "", // tester can fill in
			},
			AICheck: &scenario.AICheck{
				Prompt: "TODO: describe what should be visible here",
				Expect: "yes",
			},
		})
	}

	for _, line := range strings.Split(source, "\n") {
		line = strings.TrimSpace(line)
		if line == "" || !strings.HasPrefix(line, "await ") {
			continue
		}

		if m := rGoto.FindStringSubmatch(line); m != nil {
			url := m[1]
			if baseURL == "" {
				baseURL = url
				steps = append(steps, scenario.Step{Action: "goto", URL: "/"})
			} else {
				steps = append(steps, scenario.Step{Action: "goto", URL: url})
			}
			addScreenshot("after-goto")
			continue
		}
		if m := rRolePress.FindStringSubmatch(line); m != nil {
			steps = append(steps, scenario.Step{Action: "press", Role: m[1], RoleName: m[2], Key: m[3]})
			addScreenshot("after-press")
			continue
		}
		if m := rRoleFill.FindStringSubmatch(line); m != nil {
			steps = append(steps, scenario.Step{Action: "fill", Role: m[1], RoleName: m[2], Value: m[3]})
			continue
		}
		if m := rRoleClick.FindStringSubmatch(line); m != nil {
			steps = append(steps, scenario.Step{Action: "click", Role: m[1], RoleName: m[2]})
			addScreenshot("after-click")
			continue
		}
		if m := rLabelFill.FindStringSubmatch(line); m != nil {
			steps = append(steps, scenario.Step{Action: "fill", Label: m[1], Value: m[2]})
			continue
		}
		if m := rLabelClick.FindStringSubmatch(line); m != nil {
			steps = append(steps, scenario.Step{Action: "click", Label: m[1]})
			addScreenshot("after-click")
			continue
		}
		if m := rTextClick.FindStringSubmatch(line); m != nil {
			steps = append(steps, scenario.Step{Action: "click", Text: m[1]})
			addScreenshot("after-click")
			continue
		}
		if m := rPlaceFill.FindStringSubmatch(line); m != nil {
			steps = append(steps, scenario.Step{Action: "fill", Selector: fmt.Sprintf(`[placeholder="%s"]`, m[1]), Value: m[2]})
			continue
		}
		if m := rLocFill.FindStringSubmatch(line); m != nil {
			steps = append(steps, scenario.Step{Action: "fill", Selector: m[1], Value: m[2]})
			continue
		}
		if m := rLocClick.FindStringSubmatch(line); m != nil {
			steps = append(steps, scenario.Step{Action: "click", Selector: m[1]})
			addScreenshot("after-click")
			continue
		}
		if m := rLocSelect.FindStringSubmatch(line); m != nil {
			steps = append(steps, scenario.Step{Action: "select", Selector: m[1], Value: m[2]})
			continue
		}
	}
	return baseURL, steps
}
